AWS: Add sequential allocator for device names.

jsafrane opened the pull request:
On AWS, we should not reuse device names as long as possible, see https://aws.amazon.com/premiumsupport/knowledge-center/ebs-stuck-attaching/

> "If you specify a device name that is not in use by EC2, but is being used by the block device driver within the EC2 instance, the attachment of the EBS volume does not succeed and the EBS volume is stuck in the attaching state."

This patch adds a device name allocator that tries to find a name that's next to the last used device name instead of using the first available one. This way we will loop through all device names ("xvdba" .. "xvdzz") before a device name is reused.

Fixes: #31891

@wongma7, @gnufied, @childsb PTAL

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38818)
<!-- Reviewable:end -->

wongma7 commented:
lgtm

jsafrane committed:
AWS: Add sequential allocator for device names.

On AWS, we should not reuse device names as long as possible, see
https://aws.amazon.com/premiumsupport/knowledge-center/ebs-stuck-attaching/

"If you specify a device name that is not in use by EC2, but is being used by
the block device driver within the EC2 instance, the attachment of the EBS
volume does not succeed and the EBS volume is stuck in the attaching state."

This patch adds a device name allocator that tries to find a name that's next
to the last used device name instead of using the first available one.
This way we will loop through all device names ("xvdba" .. "xvdzz") before
a device name is reused.

jsafrane commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -365,6 +365,9 @@ type Cloud struct {
 	attaching      map[types.NodeName]map[mountDevice]awsVolumeID
+
+	// state of our device allocator
+	deviceAllocator DeviceAllocator
```
fixed, now every node has its own distinct allocator.

gnufied replied:
Will this leak as nodes get added and removed but DeviceAllocators never get cleaned up?

chrislovecnm commented:
@justinsb / @kubernetes/sig-aws / @kubernetes/sig-storage what is the priority on this??

Can we get this into 1.5.2??

cc: @sharms does this help with your problems?  Have you tested?

chrislovecnm commented:
@k8s-bot verify test this

chrislovecnm commented:
@saad-ali who owns 1.5.2?

gnufied commented:
lgtm

saad-ali commented:
> @saad-ali who owns 1.5.2?

It's me in the interim until a a 1.5 patch release czar has been selected.
Please create a cherry pick to the 1.5 branch, mark it with the 1.5 milestone and put a cherypick-candidate label on it.

Unless you believe this issues is P0, the plan right now is for 1.5.2 to be cut after the holidays (we want to avoid releases during the holiday, if possible).

hanikesn commented:
Any change to get this also in 1.4.8?

sachincab commented:
This fix would be very helpful for EBS volume attach/de-attach issues, thanks!

whereisaaron commented:
This patch and is pragmatic way to mitigate this problem. It is not honestly a 'fix'; when device name 'bb' gets tainted by being detached when still in use, this patch just puts off the time when the next EBS attach attempt is to 'bb' and fails in the 'stuck attaching' state. Basically, by doing this we just hope that before a tainted device name is reused, the node gets rebooted :-)

This excellent insight by @jingxu97 suggest that k8s is actually aware when it cause a devices name to get into this bad state:

> in k8s, when a pod is deleted, kubelet's volume controller will try to unmount first and then detach after unmount successfully. In case that there are still some mount referencing the same device, unmount may not be able to issued. The volume controller at master will timeout after a period of time (set to 6 mins), and then issue detach.

Given this I thought it would be a better idea for a node to consider 'tainted' and never try to reuse a device name after an unmount has failed (timed out) and k8s does a detach anyway. That should mean the 'stuck attaching' situation would never happen, which is one step better. Eventually a node would run out of serviceable device names, but at least the 'stuck attaching' state could be eliminated.

I make that suggestion without actually offering a better patch, so to be clear, I still fully support this patch. It does no harm, and is the best bird in hand :-)

wongma7 commented:
@whereisaaron you're right about this being only mitigation, not a proper fix. But, nobody is sure yet that this bug is caused by kubernetes not properly unmounting the device before detaching. We would need controller+kubelet logs to verify that. Because it's also possible the issue is on AWS side, e.g. maybe even after it reports a successful detach, for a short time the device is actually still in use but kubernetes will unknowingly attempt to reuse the path

jsravn commented:
re. @wongma7's comment: Every case I've investigated so far involving EBS stuck (including raising support tickets with AWS) have been due to K8s not unmounting properly. This is over the last 12 months of debugging/fixing AWS issues. There is little evidence in my mind there is any "AWS" problem apart from the fact that AWS gets into a bad state if the device isn't unmounted safely first. Similarly, force detaching like k8s now does will almost always cause the kernel state around devices to get corrupted - usually requiring a reboot of the node. I suppose this patch is a workaround to mitigate the k8s state tracking problems. Unless someone substantiates otherwise - that is, the burden of proof is to show AWS has a problem - I believe we should always assume k8s has a bug around EBSs (I used to have the opposite view, but debugging numerous issues, over a dozen EBS problems with k8s has changed that :)).
